synapse/lib/blobfile.py

```diff
@@ -0,0 +1,310 @@
+        '''
+        return self._size
+
+    def walk(self, callback=None):
+        '''
+        Walk the blobfile from the first record until the end.
+
+        Args:
+            callback (func): A callback function which is executed when a
+            BlobFile record is encountered.  The callback should accept the
+            following function signature: ``(fd, baseoff, off, size)``.  fd is
+            the blob's file descriptor; baseoff is the offset where the blob
+            header is located, offset is the location of the blob data, and
+            size is the number of bytes allocated for the blob. If no callback
+            is provided, nothing will happen when a BlobFile record is
+            encountered.
+
+        Notes:
+            In addition to the callback, when walking the blobfile additional
+            conditions may be encountered. These will fire an event and exit
+            the walk() operation.
+
+            These events will not serialize since they contain a
+            reference to the underlying file descriptor and are designed in
+            order to allow creation of tools which can inspect/consume bytes
+            from a BlobFile.
+
+                blob:walk:unpkerr - Fired if there is an error unpackting the Blobfile header.
+
+                blob:walk:truncated - Fired if there is a truncated file encountered.
+
+                blob:walk:done - Fired when the BlobWalker is done walking the file.
+
+        Returns:
+            None
+        '''
+        # Snag the max filesize
+        self.fd.seek(0, os.SEEK_END)
+        maxsize = self.fd.tell()
+
+        # Start walking at zero
+        self.fd.seek(0)
+        while True:
+
+            # Store the current blobs offset value
+            baseoff = self.fd.tell()
+
+            # Read the header
+            try:
+                header = self.fd.read(headsize)
+                size, = struct.unpack(headfmt, header)
+            except Exception as e:
+                log.exception('Failed to read/unpack header')
+                self.fire('blob:walk:unpkerr',
+                          fd=self.fd,
+                          baseoff=baseoff,
+                          size=headsize,
+                          excinfo=s_common.excinfo(e))
+                break
+
+            # Save the current fd position
+            off = self.fd.tell()
+
+            if callback:
+                callback(fd=self.fd, baseoff=baseoff, off=off, size=size)
+
+            # Calcuate the next expected header
+            next_header = off + size
+            self.fd.seek(next_header)
+
+            if next_header > maxsize:
+                self.fire('blob:walk:truncated',
+                          fd=self.fd,
+                          maxsize=maxsize,
+                          next_header=next_header)
+                break
+
+            if next_header == maxsize:
+                break
+
+        self.fire('blob:walk:done',
```

Comment:
lets kill off the blob:walk events, as their use would be dangerous anyway ( since there could be multiple walkers at the same time ) and i'm not sure i see a use case.  Thoughts?